pkg/policy/mapstate.go

```diff
@@ -117,68 +143,52 @@ type Identities interface {
-	// are.
-	len int
+	// and stores the identities in an associated builtin map.
+	trie bitlpm.Trie[bitlpm.Key[Key], map[identity.NumericIdentity]struct{}]
+	// CIDR index to keys in 'trie', one for each traffic direction
+	cidrMaps [2]map[identity.NumericIdentity]netipPrefixRefCount
+	// keys with zero identity are not in cidr, index them separately
+	anys map[Key]struct{}
 }
 
 func (msm *mapStateMap) Lookup(k Key) (MapStateEntry, bool) {
-	idMap, ok := msm.trie.ExactLookup(k.PrefixLength(), k)
-	if !ok || idMap == nil {
-		return MapStateEntry{}, false
-	}
-	v, ok := idMap[identity.NumericIdentity(k.Identity)]
+	v, ok := msm.entries[k]
 	return v, ok
 }
 
-func (msm *mapStateMap) Upsert(k Key, e MapStateEntry) {
-	var exists = true
-	idMap, ok := msm.trie.ExactLookup(k.PrefixLength(), k)
-	if !ok || idMap == nil {
-		exists = false
-		idMap = make(map[identity.NumericIdentity]MapStateEntry)
-	}
-	if _, ok := idMap[identity.NumericIdentity(k.Identity)]; !ok {
-		msm.len++
-	}
-	idMap[identity.NumericIdentity(k.Identity)] = e
-	// There is no need to do an Upsert if the
-	// map already exists.
-	if !exists {
-		kCpy := k
-		kCpy.Identity = 0
-		msm.trie.Upsert(k.PrefixLength(), kCpy, idMap)
-	}
-}
+var ip4ZeroPrefix = netip.MustParsePrefix("0.0.0.0/0")
+var ip6ZeroPrefix = netip.MustParsePrefix("::/0")
 
-func (msm *mapStateMap) Delete(k Key) (deleted bool) {
-	idMap, ok := msm.trie.ExactLookup(k.PrefixLength(), k)
-	if ok && idMap != nil {
-		if _, ok := idMap[identity.NumericIdentity(k.Identity)]; ok {
-			delete(idMap, identity.NumericIdentity(k.Identity))
-			deleted = true
-			msm.len--
+func (msm *mapStateMap) Ancestors(k Key, f func(Key, MapStateEntry) bool) (complete bool) {
+	complete = true
+	id := identity.NumericIdentity(k.Identity)
+	msm.trie.Ancestors(k.PrefixLength(), k, func(_ uint, key bitlpm.Key[policyTypes.Key], identityMap map[identity.NumericIdentity]struct{}) bool {
+		if _, ok := identityMap[id]; ok {
+			kCpy := key.Value()
+			kCpy.Identity = k.Identity
+			entry := msm.entries[kCpy]
+			if !f(kCpy, entry) {
+				complete = false
+				return false
+			}
 		}
-	}
-	if deleted && len(idMap) == 0 {
-		msm.trie.Delete(k.PrefixLength(), k)
-	}
+		return true
+	})
 	return
 }
 
-func (msm *mapStateMap) ForEach(f func(Key, MapStateEntry) bool) (complete bool) {
+func (msm *mapStateMap) Descendants(k Key, f func(Key, MapStateEntry) bool) (complete bool) {
 	complete = true
-	msm.trie.ForEach(func(_ uint, k bitlpm.Key[Key], m map[identity.NumericIdentity]MapStateEntry) bool {
-		for id, e := range m {
-			kCpy := k.Value()
-			kCpy.Identity = uint32(id)
-			if !f(kCpy, e) {
+	id := identity.NumericIdentity(k.Identity)
+	msm.trie.Descendants(k.PrefixLength(), k, func(_ uint, key bitlpm.Key[policyTypes.Key], identityMap map[identity.NumericIdentity]struct{}) bool {
+		if _, ok := identityMap[id]; ok {
+			kCpy := key.Value()
+			kCpy.Identity = k.Identity
+			entry := msm.entries[kCpy]
+			if !f(kCpy, entry) {
```

Comment:
This is only ever calling 'f' with a key that has the same `Identity` as the input key 'k'.